src/persist/src/indexed/columnar/arrow.rs

```diff
@@ -0,0 +1,288 @@
+    writer.write(&RecordBatch::from(&records))?;
+    writer.finish()?;
+    Ok(())
+}
+
+/// Decodes a BlobUnsealedBatch from the Arrow file format.
+pub fn decode_unsealed_arrow<R: Read + Seek>(r: &mut R) -> Result<BlobUnsealedBatch, Error> {
+    let file_meta = read_file_metadata(r)?;
+    let (format, meta) =
+        decode_unsealed_inline(file_meta.schema().metadata().get(INLINE_METADATA_KEY))?;
+
+    let updates = match format {
+        ProtoBatchFormat::Unknown => return Err("unknown format".into()),
+        ProtoBatchFormat::ArrowKVTD => decode_arrow_kvtd(r, file_meta)?,
+        ProtoBatchFormat::ParquetKVTD => {
+            return Err("ParquetKVTD format not supported in parquet".into())
+        }
+    };
+
+    let ret = BlobUnsealedBatch {
+        desc: SeqNo(meta.seqno_lower)..SeqNo(meta.seqno_upper),
+        updates,
+    };
+    ret.validate()?;
+    Ok(ret)
+}
+
+/// Decodes a BlobTraceBatch from the Arrow file format.
+pub fn decode_trace_arrow<R: Read + Seek>(r: &mut R) -> Result<BlobTraceBatch, Error> {
+    let file_meta = read_file_metadata(r)?;
+    let (format, meta) =
+        decode_trace_inline(file_meta.schema().metadata().get(INLINE_METADATA_KEY))?;
+
+    let updates = match format {
+        ProtoBatchFormat::Unknown => return Err("unknown format".into()),
+        ProtoBatchFormat::ArrowKVTD => decode_arrow_kvtd(r, file_meta)?,
+        ProtoBatchFormat::ParquetKVTD => {
+            return Err("ParquetKVTD format not supported in parquet".into())
+        }
+    };
+
+    let updates = updates
+        .iter()
+        .flat_map(|x| {
+            x.iter()
+                .map(|((k, v), t, d)| ((k.to_vec(), v.to_vec()), t, d))
+        })
+        .collect();
+
+    let ret = BlobTraceBatch {
+        desc: meta.desc.into_option().map_or_else(
+            || {
+                Description::new(
+                    Antichain::from_elem(u64::minimum()),
+                    Antichain::from_elem(u64::minimum()),
+                    Antichain::from_elem(u64::minimum()),
+                )
+            },
+            |x| x.into(),
+        ),
+        updates,
+    };
+    ret.validate()?;
+    Ok(ret)
+}
+
+fn decode_arrow_kvtd<R: Read + Seek>(
+    r: &mut R,
+    file_meta: FileMetadata,
+) -> Result<Vec<ColumnarRecords>, Error> {
+    let projection = None;
+    let file_reader = FileReader::new(r, file_meta, projection);
+
+    let file_schema = file_reader.schema().fields().as_slice();
+    // We're not trying to accept any sort of user created data, so be strict.
+    if file_schema != SCHEMA_ARROW_KVTD.fields() {
+        return Err(format!(
+            "expected arrow schema {:?} got: {:?}",
+            SCHEMA_ARROW_KVTD.fields(),
+            file_schema
+        )
+        .into());
+    }
+
+    let mut ret = Vec::new();
+    for batch in file_reader {
+        ret.push(ColumnarRecords::try_from(batch?)?);
+    }
+    Ok(ret)
+}
+
+impl From<&ColumnarRecords> for RecordBatch {
```

Comment:
nit: These impls are shared between the arrow and parquet encoding modules. It might make sense to move them to a common/util module.

Reply:
Either should be fine, I was just confused when I read `parquet.rs` because I didn't realize it was using the `From`/`TryFrom` impls from `arrow.rs`.

Reply:
Done